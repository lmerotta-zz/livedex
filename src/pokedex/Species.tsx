import React from "react";
import { useResource } from "rest-hooks";
import SpeciesResource from "./resources/Species";
import {
  Card,
  CardContent,
  CardTitle,
  CardSubtitle,
  CardMedia,
  CardSection
} from "../ui";
import VarietyResource from "./resources/Variety";
import Type from "./Type";

type SpeciesProps = {
  name: string;
};

const Species: React.FC<SpeciesProps> = ({ name }) => {
  const species = useResource(SpeciesResource.detailShape(), { name });
  const variety = useResource(VarietyResource.detailShape(), {
    name: species.defaultVariety.name
  });

  return (
    <Card>
      <CardContent>
        <CardMedia>
          <img src={variety.sprites.front_default} alt={variety.name} />
        </CardMedia>
        <CardTitle>
          {species.getEntryNumber()} - {species.getName()}
        </CardTitle>
        <CardSubtitle>{species.getGenus()}</CardSubtitle>
        <CardSection>
          {variety.getTypes().map(t => (
            <Type key={`${variety.name}-${t.name}`} name={t.name} />
          ))}
        </CardSection>
      </CardContent>
    </Card>
  );
};

export default Species;
