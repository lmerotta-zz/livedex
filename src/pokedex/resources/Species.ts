import { Resource } from "rest-hooks";
import { LanguageType, PokedexEntryType, NamedApiResource } from "./types";

export default class SpeciesResource extends Resource {
  readonly name: string | undefined = undefined;
  readonly names: Array<{ language: LanguageType; name: string }> = [];
  readonly pokedex_numbers: Array<{
    entry_number: number;
    pokedex: PokedexEntryType;
  }> = [];
  readonly genera: Array<{ genus: string; language: LanguageType }> = [];
  readonly varieties: Array<{
    is_default: boolean;
    pokemon: NamedApiResource;
  }> = [];

  pk() {
    return this.name;
  }

  getName(locale = "fr") {
    return this.names.find(name => name.language.name === locale)?.name;
  }

  getEntryNumber(type = "national") {
    return this.pokedex_numbers.find(entry => entry.pokedex.name === type)
      ?.entry_number;
  }

  getGenus(locale = "fr") {
    return this.genera.find(entry => entry.language.name === locale)?.genus;
  }

  get defaultVariety() {
    return this.varieties.find(v => v.is_default)!.pokemon;
  }

  static listShape<T extends typeof Resource>(this: T) {
    return {
      ...super.listShape(),
      schema: {
        results: [this.getEntitySchema()],
        count: 0,
        next: "",
        previous: ""
      }
    };
  }

  static urlRoot = "https://pokeapi.co/api/v2/pokemon-species/";
}
