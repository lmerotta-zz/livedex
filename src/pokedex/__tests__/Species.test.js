import React from "react";
import { render } from "@testing-library/react";
import { MockProvider } from "@rest-hooks/test";
import SpeciesResource from "../resources/Species";
import VarietyResource from "../resources/Variety";
import Species from "../Species";

jest.mock("../Type", () => ({ name }) => <div>{name}</div>);

const fixtures = [
  {
    request: SpeciesResource.detailShape(),
    params: { name: "bulbasaur" },
    result: {
      name: "bulbasaur",
      names: [{ language: { name: "fr" }, name: "Bulbizarre" }],
      pokedex_numbers: [{ pokedex: { name: "national" }, entry_number: 1 }],
      genera: [{ language: { name: "fr" }, genus: "Pokémon Plante" }],
      varieties: [{ pokemon: { name: "bulbasaur2" }, is_default: true }]
    }
  },
  {
    request: VarietyResource.detailShape(),
    params: { name: "bulbasaur2" },
    result: {
      name: "bulbasaur2",
      types: [
        {
          type: { name: "plant" },
          slot: 2
        },
        {
          type: { name: "poison" },
          slot: 1
        }
      ]
    }
  }
];

describe("Species unit tests", () => {
  it("Renders correctly", () => {
    const tree = (
      <MockProvider results={fixtures}>
        <Species name="bulbasaur" />
      </MockProvider>
    );
    const { queryByText, asFragment } = render(tree);

    const name = queryByText("Bulbizarre");
    expect(name).toBeDefined();

    const genus = queryByText("Pokémon Plange");
    expect(genus).toBeDefined();

    expect(asFragment()).toMatchSnapshot();
  });
});
