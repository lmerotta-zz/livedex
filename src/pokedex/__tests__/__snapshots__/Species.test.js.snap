// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Species unit tests Renders correctly 1`] = `
<DocumentFragment>
  <div
    class="w-1/2 sm:w-1/3 md:w-1/4 lg:w-1/5 xl:w-1/6 px-2 my-2"
  >
    <div
      class="flex flex-col shadow-md rounded px-4 py-2 hover:shadow-xl"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="h-24 w-24 self-center mb-2"
        >
          <img
            alt="bulbasaur2"
            src=""
          />
        </div>
        <h4
          class="text-sm leading-normal break-all"
        >
          1 - Bulbizarre
        </h4>
        <p
          class="text-xs md:text-sm text-gray-400 leading-none"
        >
          Pok√©mon Plante
        </p>
        <div
          class="my-3"
        >
          <div>
            poison
          </div>
          <div>
            plant
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
